daml-script/daml/Daml/Script.daml

```diff
@@ -156,8 +157,9 @@ data ScriptF a
 -- (try x catch f) >>= f should not pull `f` in the try block
 data CatchPayload a = CatchPayload
   with
-    act : () -> Free ScriptF (a, ())
-    handle : AnyException -> Optional a
+    act : () -> Free ScriptF (Free ScriptF LedgerValue, ())
+    handle : AnyException -> Optional (Free ScriptF LedgerValue)
+    continue : LedgerValue -> a
```

Comment:
The type argument `a` no longer appears in fields `act` and `handle`; now it's been replaced by `Free ScriptF LedgerValue`. This matches the type when used inside `Script x`, namely `a ~ Free ScriptF x`, except `x` is wrapped into a `LedgerValue`. 

This means that `instance Functor CatchPayload` now only maps over the new field `continue`, which is only evaluated and run after `act` (and after `handle` if `act` threw an exception). The old behaviour, namely mapping over `act` and `handle` was the root of the problem; it meant that all actions after the try-catch block were added to both fields, and if one of the actions was a bare `error`/`fail`/`abort`, the outer `Free ScriptF` in `act` would crash while trying to construct the inner `Free ScriptF`. This last part explains why the bug didn't happen when using `throw`, since the inner `Free ScriptF` could be constructed with a `ScriptF.Throw` to be evaluated in the following runner iteration.

Reply:
Another fix would have been to change the type of `act`,

```diff
- act : () -> Free ScriptF (a, ())
+ act : () -> Free ScriptF (() -> a, ())
```

recall that inside `Script x` `a ~ Free ScriptF x`, giving us

```diff
- act : () -> Free ScriptF (Free ScriptF x, ())
+ act : () -> Free ScriptF (() -> Free ScriptF x, ())
```

with that change, the inner computation in `act` doesn't need to be constructed until applied to `()`, so the exception cannot be caught. However, that approach would still duplicate all the actions after the try-catch block into both `act` and `handle`, so i opted against it.